### PR TITLE
[node] Allow setting encoding for Dirent readdir

### PR DESCRIPTION
**Pull request by MartyO256**
The `withFileTypes` option for [`fs.readdir`](https://nodejs.org/api/fs.html#fs_fs_readdir_path_options_callback), [`fs.readdirSync`](https://nodejs.org/api/fs.html#fs_fs_readdirsync_path_options) and [`fs.promises.readdir`](https://nodejs.org/api/fs.html#fs_fspromises_readdir_path_options) added as of v10.11.0 can be used with the `encoding` option to change the encoding of the resulting [Dirent names](https://nodejs.org/api/fs.html#fs_dirent_name).
This pull request updates the types for `fs.readdir` and `fs.readdirSync` for v10, v11 and v12 to align with the update made to `fs.promise.readdir` in https://github.com/DefinitelyTyped/DefinitelyTyped/pull/35213.

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: 
https://nodejs.org/docs/latest-v10.x/api/fs.html#fs_fs_readdir_path_options_callback
https://nodejs.org/docs/latest-v10.x/api/fs.html#fs_fs_readdirsync_path_options
- [ ] Increase the version number in the header if appropriate.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

### CODE REVIEW

File: types/node/fs.d.ts [hunk 1]
```diff
@@ -858,7 +858,7 @@ declare module "fs" {
      * @param path A path to a file. If a URL is provided, it must use the `file:` protocol.
      * @param options If called with `withFileTypes: true` the result data will be an array of Dirent.
      */
-    function readdir(path: PathLike, options: { withFileTypes: true }, callback: (err: NodeJS.ErrnoException | null, files: Dirent[]) => void): void;
+    function readdir(path: PathLike, options: { encoding?: string | null; withFileTypes: true }, callback: (err: NodeJS.ErrnoException | null, files: Dirent[]) => void): void;
```
Comment (islishude) [inline types/node/fs.d.ts]: According to the document，`encoding` has a default value `utf8` and withFileTypes has a default value `false`

So I think here should be `{ encoding?: string, withFileTypes?: boolean } `

File: types/node/fs.d.ts [hunk 2]
```diff
@@ -913,11 +913,11 @@ declare module "fs" {
     function readdirSync(path: PathLike, options?: { encoding?: string | null; withFileTypes?: false } | string | null): string[] | Buffer[];
 
     /**
-     * Asynchronous readdir(3) - read a directory.
+     * Synchronous readdir(3) - read a directory.
      * @param path A path to a file. If a URL is provided, it must use the `file:` protocol.
      * @param options If called with `withFileTypes: true` the result data will be an array of Dirent.
      */
-    function readdirSync(path: PathLike, options: { withFileTypes: true }): Dirent[];
+    function readdirSync(path: PathLike, options: { encoding?: string | null; withFileTypes: true }): Dirent[];
```
Comment (islishude) [inline types/node/fs.d.ts]: see the above.

**Review by islishude — CHANGES_REQUESTED** [review]

File: types/node/fs.d.ts (hunk 1 above)
Comment (MartyO256) [reply types/node/fs.d.ts]: Overloaded signatures were already provided to effectively support `{ encoding?: string, withFileTypes?: boolean }`.
https://github.com/DefinitelyTyped/DefinitelyTyped/blob/e8c7805c47489ed1b20821d9c22fae10be014e75/types/node/fs.d.ts#L821-L861

Comment (islishude) [reply types/node/fs.d.ts]: ok,you are right.and please put tests for it.

Comment (islishude) [reply types/node/fs.d.ts]: ok,I will approve it.

**Review by islishude — APPROVED** [review]

**Review by Flarna — CHANGES_REQUESTED** [review]
I think the `__promisify__` version of `readdir` should be also updated (line 891 as of writing this).

**Review by Flarna — APPROVED** [review]

**Comment by rbuckton** [comment]
Thanks for the contribution!

**Pull request merged**